docs/en/toolkit/env_config_ref.md

```diff
@@ -0,0 +1,342 @@
+
+## How Toolkit determines the current environment
+
+Toolkit uses a core hook called [pick_environment](https://github.com/shotgunsoftware/tk-core/blob/master/hooks/pick_environment.py) to determine which environment file to use at a given time. The `pick_environment` hook’s return values correspond to environment configuration files. For example, if `pick_environment` returns `shot_step`, Toolkit will use `config/env/shot_step.yml` to configure the Toolkit environment.
+
+## Custom environments
+
+The environment configuration files listed above are the ones provided with the Default Configuration. However, some studios may want to employ different or additional environments. For example, a studio might want different configuration settings for every stage in the pipeline – `asset_step_rig`, `asset_step_model`, `shot_step_anim`, `shot_step_light`, and so on. Fortunately, you can fully customize the available environments. 
+
+To do so, you would add the desired environment config files to the `config/env` directory, and override the `pick_environment` core hook, adding to it the logic that defines when to use your new environments. 
+
+## Basic structure
+
+Toolkit’s configuration files are written in [YAML](https://yaml.org/). The general configuration structure for any bundle (a *bundle* is an app, engine, or framework) is as follows:
+
+```
+bundle_name:
+  setting1: value
+  setting2: value
+  complex_setting:
+    sub_setting1: value  
+    Sub_setting2: value
+  location: 
+    type: descriptor_type
+    descriptor_setting1: value
+    descriptor_setting2: value
+```
+
+Bundles can be nested; the value of a setting for one bundle may itself be another bundle. Specifically, as you’ll see in the example below, every engine has an `apps` setting, and its value is the list of apps that are defined for that engine, along with each of the app’s own settings.
+
+Let’s start with a very simple example: an environment in which you have a single engine, with a single app defined within it. The following might be the contents of `project.yml` in this config:
+
+```
+engines:
+  tk-maya:
+    apps:
+      tk-multi-workfiles2:
+        location:
+          type: app_store
+          name: tk-multi-workfiles2
+          version: v0.11.8
+    location:
+        type: app_store
+        name: tk-maya
+        version: v0.9.4
+```
+
+## The engines block
+
+Every environment configuration file starts with an `engines` block. Nested within it are all of the engines defined for that environment.
+
+In our example, only a single engine is defined, `tk-maya`. It has two settings listed, `apps` and `location`. The `apps` setting is a list of all the apps defined for the engine, each with its own settings. In this case, only one app is defined for the engine, `tk-multi-workfiles2`. `location` is a special setting that every bundle requires.
```

Comment:
As you're referring to the example above in basic structure, should this and the following ones be sub headings?

> ... engine, `tk-multi-workfiles2`. `location` is a special ...

For readability could we insert "The" before the `location` or shift the `location` onto a new paragraph. The two code block blend a little and it read it as `tk-multi-workfiles2.location`.
Again maybe a little picky ;)

Reply:
re subheadings: I made engines, apps, location, and includes all subheadings of 'basic structure'.  good call.

re `workfiles2.location`: I changed the order of the sentences to mitigate this.